spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java

```diff
@@ -405,15 +407,25 @@ public boolean equals(Object other) {
       return false;
     }
 
-    // use only name in order to correctly invalidate Spark cache
+    // use name and effective snapshot id to support time travel
     SparkTable that = (SparkTable) other;
-    return icebergTable.name().equals(that.icebergTable.name());
+    return icebergTable.name().equals(that.icebergTable.name())
+        && Objects.equals(effectiveSnapshotId(), that.effectiveSnapshotId());
   }
 
   @Override
   public int hashCode() {
-    // use only name in order to correctly invalidate Spark cache
-    return icebergTable.name().hashCode();
+    // use name and effective snapshot id to support time travel
+    return Objects.hash(icebergTable.name(), effectiveSnapshotId());
+  }
+
+  public Long effectiveSnapshotId() {
```

Comment:
When you have `snapshotId` or `branch`, you can create it in advance. However, when you have neither, you should calculate `icebergTable.currentSnapshot()` every time as it can be changed.

Moreover, `icebergTable.currentSnapshot()` can be null(for example, at [this point](https://github.com/apache/iceberg/blob/4d34398cfd32465222f55df522fcd5a2db59c92c/spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTable.java#L53)), so we need to null check manually.

I'll update `SparkTable` and the test case a little bit.

Reply:
Please note the below statements, the `table`'s `currentSnapshot` is changed.

```java
assertThat(table).as("The SparkTable points to latest snapshot").isEqualTo(firstTagTable);

assertThat(table).as("The SparkTable points to latest snapshot").isEqualTo(secondTagTable);
```